db/python/tables/sequencing_group.py

```diff
@@ -556,3 +556,32 @@ async def get_type_numbers_for_project(self, project) -> dict[str, int]:
         """
         rows = await self.connection.fetch_all(_query, {'project': project})
         return {r['type']: r['n'] for r in rows}
+
+    async def get_type_numbers_history(
+        self, project_id: ProjectId
+    ) -> list[dict[str, Any]]:
```

Comment:
This might be negated by the above comment, but try to use stricter types when you do know the data shape. This could be a `TypedDict` or even a model.